lib/quicly.c

```diff
@@ -1896,17 +1937,30 @@ static int client_collected_extensions(ptls_t *tls, ptls_handshake_properties_t
 
     const uint8_t *src = slots[0].data.base, *end = src + slots[0].data.len;
     quicly_transport_parameters_t params;
-    quicly_cid_t odcid;
+    quicly_cid_t original_dcid, initial_scid, retry_scid = {};
 
-    /* decode and validate */
+    /* decode */
     assert(conn->super.remote.cid_set.cids[0].sequence == 0);
-    if ((ret = quicly_decode_transport_parameter_list(&params, &odcid, conn->super.remote.cid_set.cids[0].stateless_reset_token, 1,
-                                                      src, end)) != 0)
+    if ((ret = quicly_decode_transport_parameter_list(&params, &original_dcid, &initial_scid, is_retry(conn) ? &retry_scid : NULL,
+                                                      conn->super.remote.cid_set.cids[0].stateless_reset_token, 1, src, end)) != 0)
+        goto Exit;
+
+    /* validate CIDs */
+    if (!quicly_cid_is_equal(&conn->super.local.cid_set.cids[0].cid, ptls_iovec_init(original_dcid.cid, original_dcid.len))) {
```

Comment:
how are the scids and dcids organized in this cid_set? Not for this PR, but it might be easier to read/debug if the names had scid/dcid in them.

Reply:
I'm not sure if I allow.

We cannot use SCID / DCID as terms referring to states. That is because SCID / DCID are tied to the direction of the packet. A CID issued by an endpoint will appear as DCID in ingress packets, or as SCID in egress packets.

Reply:
Sorry I wasn't clear. What I'm saying is that `local.cid_set.cids[0].cid` does not tell me which CID it is, but it is supposed to be the Original DCID. Similarly, `remote.cid_set.cids[0].cid` is used for the Initial SCID,  but the name doesn't make that obvious.